### PR TITLE
[Android] Add response headers to XHR response

### PR DESCRIPTION
**Pull request by nucleartux**
This is code from https://github.com/facebook/react-native/pull/3062 with fix for support multiple values from @qbig

### CODE REVIEW

**Comment by nucleartux** [comment]
@ide @mkonicek any feedback?

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 1]
```diff
@@ -189,7 +189,22 @@ public void onResponse(Response response) throws IOException {
...
-            callback.invoke(response.code(), null, responseBody);
+
+            WritableMap responseHeaders = Arguments.createMap();
+            Headers headers = response.headers();
+            for (int i = 0; i < headers.size(); i++) {
+              // multiple values for the same header
+              if (responseHeaders.hasKey(headers.name(i))) {
+                responseHeaders.putString(
+                  headers.name(i),
+                  responseHeaders.getString(headers.name(i) + "," + headers.value(i))
```
Comment (ide) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: is using a comma the way you're supposed to do this?

Comment (nucleartux) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: Yes. HTTP RFC2616 available says: 
Multiple message-header fields with the same field-name MAY be present in a message if and only if the entire field-value for that header field is defined as a comma-separated list [i.e., #(values)]. It MUST be possible to combine the multiple header fields into one "field-name: field-value" pair, without changing the semantics of the message, by appending each subsequent field-value to the first, each separated by a comma. The order in which header fields with the same field-name are received is therefore significant to the interpretation of the combined field value, and thus a proxy MUST NOT change the order of these field values when a message is forwarded.

Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: This doesn't look correct: you're getting the string out of the responseHeaders map corresponding to the key `{header_key},{header_value}`. Was this tested?

**Comment by astreet** [comment]
Could you add an example of this working to the XHRExample? (https://github.com/facebook/react-native/blob/master/Examples/UIExplorer/XHRExample.android.js)

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java (hunk 1 above)
Comment (nucleartux) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: Oh, I will fix this to values separated by comma

**Comment by astreet** [comment]
Also, is this how iOS sends response headers as well? (i.e. do they do a comma separated list or do they send an array). We want to make sure these APIs are consistent.

**Comment by nucleartux** [comment]
Yes, this is consistent with iOS

**Comment by nucleartux** [comment]
@astreet @ide I fixed typo and added example, can you review this changes?

File: Examples/UIExplorer/XHRExample.android.js [hunk 2]
```diff
@@ -259,6 +259,85 @@ class FormUploader extends React.Component {
   }
 }
 
+class Headers extends React.Component {
```
Comment (ide) [inline Examples/UIExplorer/XHRExample.android.js]: Can you pull this out into a shared file like XHRExampleHeaders.js so it's not duplicated across platforms?

Comment (nucleartux) [inline Examples/UIExplorer/XHRExample.android.js]: done :)

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 3]
```diff
@@ -189,7 +189,22 @@ public void onResponse(Response response) throws IOException {
...
+
+            WritableMap responseHeaders = Arguments.createMap();
+            Headers headers = response.headers();
+            for (int i = 0; i < headers.size(); i++) {
+              // multiple values for the same header
+              if (responseHeaders.hasKey(headers.name(i))) {
+                responseHeaders.putString(
+                  headers.name(i),
+                  responseHeaders.getString(headers.name(i)) + ", " + headers.value(i)
+                ) ;
```
Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: put this on the previous line, no whitespace before ;

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 4]
```diff
@@ -189,7 +189,22 @@ public void onResponse(Response response) throws IOException {
               callback.invoke(0, null, e.getMessage());
               return;
             }
-            callback.invoke(response.code(), null, responseBody);
+
+            WritableMap responseHeaders = Arguments.createMap();
+            Headers headers = response.headers();
+            for (int i = 0; i < headers.size(); i++) {
+              // multiple values for the same header
+              if (responseHeaders.hasKey(headers.name(i))) {
```
Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: no reason not to cache `headers.name(i)`

**Comment by astreet** [comment]
Looks great, thanks! Can you just update the inlines to keep with our internal code formatting so I can just auto-import without having to touch it? :)

**Comment by astreet** [comment]
@facebook-github-bot shipit

**Comment by astreet** [comment]
Thanks!

**Comment by qbig** [comment]
awesome!

File: Examples/UIExplorer/XHRExampleHeaders.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+ * Facebook reserves all rights not expressly granted.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL
+ * FACEBOOK BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
+ * AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * @flow
```
Comment (astreet) [inline Examples/UIExplorer/XHRExampleHeaders.js]: This is blocking commit on our side. Please make sure this file passes flow checks (it currently doesn't) or remove the `@flow` annotation. I'll re-submit once this is updated :)

**Comment by astreet** [comment]
@facebook-github-bot shipit

**Comment by sreesharp** [comment]
Hi @nucleartux @astreet, I am working on a PR to support the "response" attribute to XHR in order to solve this issue ( https://github.com/facebook/react-native/issues/6017), and just now noticed this PR. Let me know if you have any thoughts about supporting "response" attribute, in addition to the "responseText".

**Pull request closed without merging**